lib/mshoplib/src/MShop/Order/Manager/Base/Base.php

```diff
@@ -471,12 +483,19 @@ protected function loadItems( $id, \Aimeos\MShop\Price\Item\Iface $price,
 	protected function loadFresh( $id, \Aimeos\MShop\Price\Item\Iface $price,
 		\Aimeos\MShop\Locale\Item\Iface $localeItem, $row, $parts )
 	{
-		$products = $addresses = $services = [];
+		$products = $coupons = $addresses = $services = [];
 
-		if( $parts & \Aimeos\MShop\Order\Item\Base\Base::PARTS_PRODUCT ) {
+		if( $parts & \Aimeos\MShop\Order\Item\Base\Base::PARTS_PRODUCT
+			|| $parts & \Aimeos\MShop\Order\Item\Base\Base::PARTS_COUPON
+		) {
 			$products = $this->loadProducts( $id, true );
 		}
 
+		if( $parts & \Aimeos\MShop\Order\Item\Base\Base::PARTS_COUPON ) {
+			// load coupons with product array containing product ids for coupon/product matching
+			$coupons = $this->loadCoupons( $id, true, $this->loadProducts( $id, false ) );
```

Comment:
You should pass `$products` if not empty to avoid double loading of all products

Reply:
Well, I can't, because `$products` is "fresh" (without ids) - so no coupon matching is possible. Any other suggestion than loading all products with ids again?

Reply:
Let's stay with your current code to avoid bigger changes. Don't think it will be a performance problem right now and we think about a better implementation for 2020.01